src/coreclr/jit/llvm.cpp

```diff
@@ -452,7 +640,27 @@ Value* buildInd(llvm::IRBuilder<>& builder, GenTree* node, Value* ptr)
 
 Value* buildNe(llvm::IRBuilder<>& builder, GenTree* node, Value* op1, Value* op2)
 {
-    return mapTreeIdValue(node->gtTreeID, builder.CreateICmpNE(op1, op2));
+    // TODO: when the next integer binary operator is implemented, factor out the widening
+    Type* op1Type = op1->getType();
+    Type* op2Type = op2->getType();
+    if (op1Type == op2Type)
+    {
+        // no widening required
+        return mapTreeIdValue(node->gtTreeID, builder.CreateICmpNE(op1, op2));
+    }
+    else
```

Comment:
I am curious: where are you seeing the type mismatches? The frontend should normalize them in almost all cases, safe for shifts.

Reply:
I hit this with I think the short circuit evaluation of e.g. https://github.com/dotnet/runtimelab/blob/5054e33a8bc08e05da77233ed2d2ca29a2400a56/src/coreclr/nativeaot/System.Private.CoreLib/src/System/RuntimeExceptionHelpers.cs#L239.  The IR being processed starts with
```
------------ BB01 [000..008) -> BB03 (cond), preds={} succs={BB02,BB03}
               [000076] ------------                 IL_OFFSET void   IL offset: 0x0
N001 (  1,  1) [000000] ------------                 NO_OP     void
N001 ( 14,  2) [000001] H-CXG-------         t1 =    CALL help byref  HELPER.CORINFO_HELP_READYTORUN_STATIC_BASE Zero Fseq[Value]
                                                  /--*  t1     byref
N002 ( 18,  5) [000004] ---XG-------         t4 = *  IND       bool
N003 (  1,  1) [000005] ------------         t5 =    CNS_INT   int    0
                                                  /--*  t4     bool
                                                  +--*  t5     int
N004 ( 20,  7) [000006] J--XG--N----         t6 = *  NE        int
                                                  /--*  t6     int
N005 ( 22,  9) [000007] ---XG-------              *  JTRUE     void
```
Where the NE operates on `t4` and `t5`.  `t4` is a bool which I currently translate to an `i8` in LLVM and `t5` is an `int`, LLVM `i32` .  The LLVM `icmp` won't like that.  I suppose in this case as `t5` is 0, that could go to an `i8` which would be more optimal and avoid the widening.  Maybe I need to look harder and there is more information available in `CNS_INT   int    0` but not dumped here, that tells me 8 bits is enough.

Reply:
> N002 ( 18,  5) [000004] ---XG-------         t4 = *  IND       bool

Ah, of course, the indirections off of small types. Those always implicitly extend in IR (so while in IR the type is e. g. `short`, the type that the user consumes is `int`).

I would personally go with just upcasting the indirections (and a couple other nodes that are indir-like: `CLS_VAR`, `LCL_FLD`), since that should be the only case where this happens (well, there's one more case: morph can retype relops as `TYP_UBYTE` sometimes, but I have a PR which moves this to lowering). But up to you if you want to follow the IR semantics or build on top of them here.

>  Maybe I need to look harder and there is more information available in CNS_INT int 0 but not dumped here, that tells me 8 bits is enough.

I don't think so. Constants in this phase are always `LONG` or `INT`, there are no small types for them (only codegen sometimes retypes them to bytes to emit "narrow" instructions on XARCH).

Reply:
Thanks, have gone for upcasting the indirection.